test/Serialization/Recovery/implementation-only-missing.swift

```diff
@@ -0,0 +1,85 @@
+// Recover from missing types hidden behind an importation-only when indexing
+// a system module.
+// rdar://problem/52837313
+
+// RUN: %empty-directory(%t)
+
+//// Build the private module, the public module and the client app normally.
+//// Force the public module to be system with an underlying Clang module.
+// RUN: %target-swift-frontend -emit-module -DPRIVATE_LIB %s -module-name private_lib -emit-module-path %t/private_lib.swiftmodule
+// RUN: %target-swift-frontend -emit-module -DPUBLIC_LIB %s -module-name public_lib -emit-module-path %t/public_lib.swiftmodule -I %t -I %S/Inputs/implementation-only-missing -import-underlying-module
+
+//// The client app should build OK without the private module. Removing the
+//// private module is superfluous but makes sure that it's not somehow loaded.
+// RUN: rm %t/private_lib.swiftmodule
+// RUN: %target-swift-frontend -typecheck -DCLIENT_APP -primary-file %s -I %t -index-system-modules -index-store-path %t
+
+#if PRIVATE_LIB
+
+public struct HiddenGenStruct<A: HiddenProtocol> {
+  public init() {}
+}
+
+public protocol HiddenProtocol {
+  associatedtype Value
+
+  func hiddenFunc<A>(a : HiddenGenStruct<A>)
+}
+
+#elseif PUBLIC_LIB
+
+@_implementationOnly import private_lib
+
+struct PublicTypeAliasBound { }
+
+struct S: HiddenProtocol {
+  typealias Value = PublicTypeAliasBound
+
+  func hiddenFunc<A>(a : HiddenGenStruct<A>) { }
+}
+
+protocol PublicProtocolAsBounds {
+  init<B>(b : HiddenGenStruct<B>)
+}
+
+struct LayoutContext : PublicProtocolAsBounds {
+  public init<B>(b : HiddenGenStruct<B>) { }
+}
+
+protocol PublicProtocol {
+  associatedtype TA : PublicProtocolAsBounds = LayoutContext
+
+  func hiddenRequirement<A>(
+      param: inout HiddenGenStruct<A>
+  ) where A.Value == PublicTypeAliasBound
+}
+
+extension PublicProtocol where TA == LayoutContext {
+  func hiddenRequirement<A>(
+      param: inout HiddenGenStruct<A>
+  ) where A.Value == PublicTypeAliasBound { }
+}
+
+public struct PublicStruct : PublicProtocol {
+  typealias TA = LayoutContext
+
+  public init<A>(a : A) {}
+
+  public func publicFunc() {
+    var s = HiddenGenStruct<S>()
+    hiddenRequirement(param: &s)
+  }
+
+  func hiddenRequirement<A>(
+      param: inout HiddenGenStruct<A>
+  ) where A.Value == PublicTypeAliasBound { }
+}
```

Comment:
Do we still need everything in the public library for the test to exercise the problem? I suspect a lot of this can go away or be simplified. (At the very least none of the function bodies matter, and the `inout`s seem like a distraction.)

Reply:
We probably don't need everything. I could reduce it further, but it could be useful to have a larger test since this path wasn't much tested already. What do you think?

Reply:
It's true that the path isn't tested much, but I'm not sure the larger test actually tests more of the API surface (as opposed to just testing the same thing more, or parts that don't depend on the implementation-only module at all).

Reply:
I reduced it further and removed the redundant code.